[JSC] JITWorklist::visitWeakReferences() should bail if the VM has no active plans

dhecht opened the pull request:
#### d8d0d118bb8b13b08d72d5ee92b7391a41c8cdf6
<pre>
[JSC] JITWorklist::visitWeakReferences() should bail if the VM has no active plans
<a href="https://bugs.webkit.org/show_bug.cgi?id=293711">https://bugs.webkit.org/show_bug.cgi?id=293711</a>
<a href="https://rdar.apple.com/150606516">rdar://150606516</a>

Reviewed by Keith Miller and Mark Lam.

When the VM being collected has no JITPlans, there&apos;s no point in
JITWorklist::visitWeakReferences() searching for a plan, so bail
earlier like the other methods do. This avoids a racy check
of the safepoint since the JIT threads might not be suspended.
Add an assert that verifies the invariant that avoids this race.

Canonical link: <a href="https://commits.webkit.org/295564@main">https://commits.webkit.org/295564@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1e32add00087e70b38acf61fa46378cacdc1e45

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25162 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15590 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110652 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56076 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25622 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33701 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80090 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108457 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19997 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95169 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60399 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19728 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13257 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55489 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/98097 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13298 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113402 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/104069 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32594 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24037 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89169 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91396 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88824 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22654 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33696 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11498 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28054 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32520 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37945 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/128370 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32277 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35089 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35624 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33863 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/jit/JITWorklist.cpp:
```diff
@@ -314,6 +314,7 @@ void JITWorklist::removeDeadPlans(VM& vm)
     });
 
     // No locking needed for this part, see comment in visitWeakReferences().
+    ASSERT(m_suspensionLock.isLocked());
```
I would do this with `WTF_REQUIRES_LOCK(m_suspensionLock)` on the declaration of `removeDeadPlans`

dhecht replied:
The lock is not required to call this function (without refactoring things a bit). In any case, I've rewritten this assert to be on each `m_rightToRun` lock instead, which I think is clearest.

kmiller68 commented on Source/JavaScriptCore/jit/JITWorklist.cpp:
```diff
@@ -356,6 +359,7 @@ void JITWorklist::visitWeakReferences(Visitor& visitor)
     // (1) no new threads can be added to m_threads. Hence, it is immutable and needs no locks.
     // (2) JITWorklistThread::m_safepoint is protected by that thread's m_rightToRun which we must be
     //     holding here because of a prior call to suspendAllThreads().
+    ASSERT(m_suspensionLock.isLocked());
```
Ditto on `WTF_REQUIRES_LOCK(m_suspensionLock)` but for `visitWeakReferences`

kmiller68 reviewed (approved):
r=me too.

dhecht committed:
[JSC] JITWorklist::visitWeakReferences() should bail if the VM has no active plans
https://bugs.webkit.org/show_bug.cgi?id=293711
rdar://150606516

Reviewed by Keith Miller and Mark Lam.

When the VM being collected has no JITPlans, there's no point in
JITWorklist::visitWeakReferences() searching for a plan, so bail
earlier like the other methods do. This avoids a racy check
of the safepoint since the JIT threads might not be suspended.
Add an assert that verifies the invariant that avoids this race.

Canonical link: https://commits.webkit.org/295564@main
